File: param_tutorial/src/calculate_server.cpp
#include "ros/ros.h"
#include "param_tutorial/Calculate.h"

#define PLUS 1            // define == 매크로 상수
#define MINUS 2
#define MULTIPLICATION 3
#define DIVISION 4


int my_operator = MINUS; //전역 변수로 쓴 이유, main에서도 써야되고 스위치문에서도 써야되서.
bool calculation(param_tutorial::Calculate::Request& req,
                 param_tutorial::Calculate::Response& res)

{
    switch (my_operator)                 //
    {
        case PLUS:
            res.result = req.a + req.b;
            break;
         case MINUS:
            res.result = req.a - req.b;
            break;
         case MULTIPLICATION:
            res.result = req.a * req.b;
            break;
         case DIVISION:
            res.result = req.a / req.b;
            break;
        default:
            res.result = req.a + req.b;
            break;            
    }
    ROS_INFO("a: %d, b: %d, result: %d", req.a, req.b, res.result);

    return true;
}

int main(int argc, char** argv)
{
    ros::init(argc,argv,"calculate_server");
    ros::NodeHandle nh;

    nh.setParam("calculation_method", MINUS); // set(설정해준다)이라는 함수는 파라미터 서버에 calculation_method 라는 이름의 변수안에 PLUS(1)를 넣어준다.

    ros::ServiceServer calculate_server = nh.advertiseService("calculate",calculation);

    ros::Rate loop_late(10);
    while (ros::ok())
    {
        nh.getParam("calculation_method", my_operator);
        //파라미터 서버에 있는 calculation_method 라는 이름의 변수를 my_operator 여기로 가져와서 할당 저장하겠다.
        ros::spinOnce();
         //스핀을 한번만 해주는 함수(콜백 한번하고 끝). ros::spin()-계속 대기하면서 콜백함수를 호출해주는것. 리퀘스트가 오면 계속 내어준다.
        loop_late.sleep();
    }
    
    
    return 0;
}

File: param_tutorial/scripts/calculate_server.py
#!/usr/bin/python3
#-*- coding: utf-8 -*- 


import rospy
from param_tutorial.srv import Calculate, CalculateResponse

PLUS = 1 
MINUS = 2
MULTIPLICATION = 3
DIVISION = 4

def calculation(req):
    my_operator = rospy.get_param("calculation_method")
    if my_operator == PLUS:
        result = req.a + req.b
    elif my_operator == MINUS:
        result = req.a - req.b
    elif my_operator == MULTIPLICATION:
        result = req.a * req.b 
    elif my_operator == DIVISION:
        if req.b == 0:
            result = 0
        else:
            result = req.a // req.b
    else:
        result = req.a + req.b

    rospy.loginfo(f"a: {req.a}, b: {req.b}, result: {result}")
    res = CalculateResponse()
    res.result = result
    return res

def calculate_server():
    rospy.init_node("calculate_server")

    rospy.set_param("calculation_method", PLUS)
    my_server = rospy.Service("calculate", Calculate, calculation ) #서버이름, srv이름,함수이름

    rospy.loginfo("Service Server ready")
    rospy.spin()

if __name__ == "__main__":
    calculate_server()
